Docs: How to interface external codes (#4093)
aiidateam/aiida-core: docs/source/howto/codes.rst

.. code-block:: python

    spec.input('code', valid_type=orm.Code, help='The `Code` to use for this job.')

Moreover, as this information is necessary for the engine, this input has a ``required=True`` setting (the default for all inputs, including the ones we manually defined earlier).
For the plugin user this means they will have to provide it when instantiating the calculation (as can be seen in the :ref:`respective section<how-to:codes:run>` below), whereas for the plugin developer this means they will have to make sure to manually pass its UUID from the ``self.inputs['code']`` node to the ``codeinfo.code_uuid`` property, as shown here.

The other two lines are configuring how to build the running script: line 14 indicates where to redirect the standard output, whereas line 15 lists the command line arguments to be passed to the code.
The specific combination presented here, together with the information inside of the |Code| node, results in the following script:

.. code-block:: bash

    #!/bin/bash

>     '[executable path in code node]' '[input_filename]' > '[output_filename]'

Reviewer: ```suggestion
    '[executable path in code node]' < '[input_filename]' > '[output_filename]'
```

You can set a parser as the default option in the |define| method, but note that this choice can be overriden when instantiating the |CalcJob|.

.. code-block:: python

    @classmethod
    def define(cls, spec):
        (...)
        spec.inputs['metadata']['options']['parser_name'].default = 'arithmetic.add'

As can be seen in the previous line, the way to do this is not by passing the |Parser| class directly, but by providing the string label that identifies the |Parser| as a registered plugin in your working environment.
In other words, in order to use a |Parser| you will need to register it as explained in the following how-to section on :ref:`registering plugins <how-to:plugins>`.

Handling parsing errors
-----------------------

In order for you to be able to provide the user with information regarding the errors that can ocur after the calculation has finished (so, mostly during parsing), you have the option of defining ``exit_codes`` for cases when something goes wrong.

> The typical way to implement these is to use ``try``/``except`` clauses to wrap the lines of code that might raise some typical errors, that can then be replaced by these ``exit_codes``:

Reviewer: I would not frame it as try/except because that is not all necessary or the idea behind. There don't need to be exceptions for one to want to return an exit code. For example you can check the output of the file, and all works just fine, but you conclude the output is not as desired, so you return an exit code. The try/except is not at all linked to in such a direct way. I would rather propose something like the following:
```suggestion
To signal a particular problem, you should just return the corresponding exit code.
These can be easily referenced in the parser through the ``self.exit_codes`` property.
For example, if the |CalcJob| defines an exit code with the label ``ERROR_READING_OUTPUT_FILE``, you can retrieve it as ``self.exit_codes.ERROR_READING_OUTPUT_FILE``.
If an exit code is returned from the parser, the engine will set it on the corresponding calculation job node.
```

.. code-block:: python
   :linenos:

    def parse(self, **kwargs):
        from aiida.common import exceptions

        try:
            output_folder = self.retrieved
        except exceptions.NotExistent:
            return self.exit_codes.ERROR_NO_RETRIEVED_FOLDER

        output_filename = self.node.get_attribute('output_filename')

        try:
            with output_folder.open(output_filename, 'r') as handle:
                try:
                    result = int(handle.read())
                except ValueError:

>                     result = None

Reviewer: ```suggestion
                    return self.exit_codes.ERROR_INVALID_OUTPUT
```

>         except (OSError, IOError):

Reviewer: ```suggestion
        except OSError:
```
In Python 3 we can safely only catch `OSError`

            return self.exit_codes.ERROR_READING_OUTPUT_FILE

>         if result is None:
>             return self.exit_codes.ERROR_INVALID_OUTPUT

Reviewer: ```suggestion
```

        self.out('sum', orm.Int(result))

You can see there is one for the case where no ``retrieve`` output was found (lines 4-7), another for the case of not being able to read the output file (lines 11-18), and finally one for when the result printed in the file is not a valid integer (lines 20-21).
You then have to introduce all of these `exit_codes` inside of the |define| method of the |CalcJob| that will use this |Parser| (so, in that sense, when working like this the exit codes need to be "supported" by any calculation that wants to use the |Parser|).

.. code-block:: python

    @classmethod
    def define(cls, spec):
        (...)
        spec.exit_code(300, 'ERROR_NO_RETRIEVED_FOLDER', message='The retrieved folder data node could not be accessed.')
        spec.exit_code(310, 'ERROR_READING_OUTPUT_FILE', message='The output file could not be read from the retrieved folder.')
        spec.exit_code(320, 'ERROR_INVALID_OUTPUT', message='The output file contains invalid output.')

> As you can see, for each ``exit_code`` you need to provide an ID number that will be used to identify it, a label you can then use to reference the code in the |parse| method (``self.exit_codes.LABEL``), and a message that will give the user more information on the problem.

Reviewer: ```suggestion
As you can see, for each ``exit_code`` you need to provide a exit status (which should be a positive integer), a label that can be used to reference the code in the |parse| method (``self.exit_codes.LABEL``), and a message that provides a more detailed information on the problem.
```

> The Topics section on :ref:`defining processes <topics:processes:usage:defining>` also contains more information on how to use exit codes.

Reviewer: ```suggestion
The Topics section on :ref:`defining processes <topics:processes:usage:defining>` also contains more information on how to use exit codes.
```
```suggestion
The Topics section on :ref:`defining processes <topics:processes:usage:defining>` provides additional information on how to use exit codes.
```

For the code that adds up two numbers, we will need to define those numbers as inputs (lets call them ``x`` and ``y`` to label them) and the result as an output (``sum``).

.. code-block:: python
   :linenos:

    @classmethod
    def define(cls, spec):
        from aiida import orm
        super().define(spec)

        spec.output('sum', valid_type=orm.Int, help='The sum of the left and right operand.')
        spec.input('x', valid_type=orm.Int, help='The left operand.')
        spec.input('y', valid_type=orm.Int, help='The right operand.')

        spec.inputs['metadata']['options']['input_filename'].default = 'aiida.in'
        spec.inputs['metadata']['options']['output_filename'].default = 'aiida.out'

The first line after the import just runs the ``define`` method of the parent base |CalcJob| class, which will do some set-ups and define some basic `inputs` and `outputs` (we will see some of them shortly).
The second "block of code" (lines 6-8), is where we are defining the specific inputs and outputs for the code, specifying their valid type (in this case AiiDA nodes of type |Int|) and a help message for the users of the class.

> The last block (lines 10-12) seems similar to the previous one, but has a subtle difference: it is not defining new `inputs`, but modifying some properties of the base `inputs` that are already defined in the parent |CalcJob| class.

Reviewer 2: ```suggestion
The last block (lines 10+11) seems similar to the previous one, but has a subtle difference: it is not defining new `inputs`, but modifying some properties of the base `inputs` that are already defined in the parent |CalcJob| class.
```

        input_code = self.inputs['code']
        input_filename = self.inputs['metadata']['options']['input_filename']
        output_filename = self.inputs['metadata']['options']['output_filename']

        with folder.open(input_filename, 'w', encoding='utf8') as handle:
            handle.write('echo $(( {} + {} ))\n'.format(input_x.value, input_y.value))

        codeinfo = CodeInfo()
        codeinfo.code_uuid = input_code.uuid
        codeinfo.stdout_name = output_filename
        codeinfo.cmdline_params = [input_filename]

        calcinfo = CalcInfo()
        calcinfo.codes_info = [codeinfo]
        calcinfo.retrieve_list = [output_filename]

        return calcinfo

The first block of code (lines 3-7) is just the unloading of the information stored in the ``spec`` into local variables.
Note that this information is not accessed via the ``spec.inputs`` anymore, but by ``self.inputs``: by the time this method is executed the specs will have become properties of the ``CalcJob`` and now should contain the actual inputs provided by the user.

The input required by the addition "code" is just a `bashscript` line with the value of the input nodes replaced appropriately.

> This is being created on line 9, using the ``folder.open`` method to get a handle to the file and simply wirting in it.

Reviewer 2: I think it would be good to link to API reference here. The same goes for many other code elements named here.

Author: I added most of them to the docs. The only ones that were left without an API reference were the `spec`, `retrieve_list`(because I couldn't find a specific useful method to link to) and one `ArithmeticAddParser` mention (because the exposed implementation is a simplified version from the current one shipped with aiida so I think it could be confusing).

Reviewer 2: Ok.

This directory represented by the |Folder| object (along with all the files created in it) will not only be copied to the remote machine for the code to be run there, but will also be stored in the local repository of the calculation node.

Next in lines 12-15 we are creating and setting up the ``codeinfo = CodeInfo()`` object.

> The ``code_uuid`` being passed in line 13 is necessary for the engine to get the base information from the |Code| node (in which computer the code is placed, what is the location of the executable, etc).

Reviewer 2: Somehow the vertical bars are showing up in the [formatted output here](https://36-77234579-gh.circle-artifacts.com/0/html/howto/codes.html).

> Note that this was taken in the unloading block from ``input_code = self.inputs['code']``, which we never specified in the ``define`` method: this is one of the inputs defined in the base |CalcJob| class that we mentioned earlier when discussing the ``super().define(spec)``:

Reviewer 2: I think ``define`` should also be linked to the API reference.

The other two lines are configuring how to build the running script: line 14 indicates where to redirect the standard output, whereas line 15 lists the command line arguments to be passed to the code.
The specific combination presented here, together with the information inside of the ``code`` node, results in the following script:

.. code-block:: bash

    #!/bin/bash

    '[executable path in code node]' '[input_filename]' > '[output_filename]'

Through the |CodeInfo| object you can also pass more flags (by adding them as string elements to the list in ``codeinfo.cmdline_params``), configure what to pass through the standard input (just as it is shown for the standard output), add commands to be run before and after the execution line, etc.

Finally, the last block remaining in lines 17-19 creates the |CalcInfo| object, passes to it the |CodeInfo| object, and adds the output to the ``retrieve_list``.

> This is a list of all files that the engine needs to copy back from the computer where the code is located, either for immediate parsing or local storage in an output node labeled ``retrieved`` for future post-processing (again, this is another base ``output`` defined in the base |CalcJob| class).

Reviewer 2: ```suggestion
This is a list of all files that the engine needs to copy back from the computer where the code is located, either for immediate parsing or local storage in an output node labeled ``retrieved`` for future post-processing.
The ``retrieved`` node is a ``output`` defined in the base |CalcJob| class.
```

> You can expand this list to include more files, and there are other lists available that allow you to easily specify which files to make available in the remote folder without having to manually copy them during this step and thus having to store unnecessary copies of them in the calculation node.

Reviewer 2: This is a *very long* sentence and I find it quite confusing. You are mentioning this and other lists (without actually specifying them) to let the user know how to control what is copied and what is not copied? I'd appreciate if you could try to reword this sentence. It may be warranted its own paragraph.

Author: Ok,, removed the clarification for expansion (it is probably self evident) and shortened a bit the phrase about the other lists:

This is a list of all files that the engine needs to copy back from the computer where the code is located, either for immediate parsing or local storage in an output node labeled ``retrieved`` for future post-processing.
The ``retrieved`` node is a ``output`` defined in the base |CalcJob| class.
There are other lists available that allow you to easily customize how to move files to and from the remote working directory in order to prevent the creation of unnecessary copies.

In the Topics section on :ref:`defining calculations <topics:calculations:usage>` you will find more information on available settings of the |CalcInfo| and |CodeInfo|, such as available copy lists, running script options, etc.

Design guidelines

> -----------------

Reviewer 2: Something is off here. Why is this followed by the "Design guidelines" completely out of context? Did something go wrong during a merge/rebase?

Author: "Design guidelines" was something that @ltalirz  put here in one of his recen PR's, I think the plugin ones. I was not sure how to incorporate them into the text of the how-to so I left it out as a final sub-section (I only separated the item that referred to parsing and moved it at the end of the respective section). I'm not really sure what to do with it other than this but I also didn't want to just delete it.

Reviewer 2: Ok, then let's just leave it there and I'll try to figure it out later.

 * | **Check existing resources.**
   | Before starting to write a plugin, check on the `aiida plugin registry <https://aiidateam.github.io/aiida-registry/>`_ whether a plugin for your code is already available.
 * | **Start simple.**
   | Make use of existing classes like :py:class:`~aiida.orm.nodes.data.dict.Dict`, :py:class:`~aiida.orm.nodes.data.singlefile.SinglefileData`, ...
   | Write only what is necessary to pass information from and to AiiDA.
 * | **Don't break data provenance.**
   | Store *at least* what is needed for full reproducibility.
 * | **Expose the full functionality.**
   | Standardization is good but don't artificially limit the power of a code you are wrapping - or your users will get frustrated.
   | If the code can do it, there should be *some* way to do it with your plugin.
 * | **Don't rely on AiiDA internals.**
   | AiiDA's :ref:`public python API<python_api_public_list>` includes anything that you can import via ``from aiida.module import thing``.
   | Functionality at deeper nesting levels is not considered part of the public API and may change between minor AiiDA releases, breaking your plugin.

.. _how-to:codes:parsing:

Parsing the outputs

> ===================

Reviewer 2: Shouldn't this be a sub-section of "Interfacing external code" ?

The parsing step occurs after the calculation has finished running and all the relevant outputs have been retrieved.
It is an optional step that allows you to extract relevant information from the output files and store it into AiiDA nodes in formats that are easier and quicker to query and analyze.
The way to implement this is by extending the base |Parser| class as you can see below:

.. code-block:: python
   :linenos:

    from aiida.parsers.parser import Parser
    from aiida import orm

    class ArithmeticAddParser(Parser):

        def parse(self, **kwargs):

            output_folder = self.retrieved
            output_filename = self.node.get_attribute('output_filename')

            with output_folder.open(output_filename, 'r') as handle:
                result = int(handle.read())

            self.out('sum', orm.Int(result))

> The first command in the example (line 8) shows the way to get the ``retrieve`` folder that was generated by the associated ``CalcJob`` and contains the files included in the ``retrieve_list`` during the ``prepare_for_submission``.

Reviewer 2: > during the ``prepare_for_submission``.

What does that mean?

Author: Changed it, clearer?:

The first command in the example (line 8) shows the way to get the ``retrieve`` folder that was generated by the associated |CalcJob| and that, by this point, should contain the files included in the ``retrieve_list`` (as specified in the |prepare_for_submission| method).

Reviewer 2: Yes, that's clearer.

The second one (line 9) is also accessing a parameter of the |CalcJob|: in this case, the name of the output.
After unloading this information into local variables, these are then used to open said output file located in the obtained retrieved folder and read the single integer value that was written there by the addition ``code`` (lines 11 and 12).

Finally, it uses the ``out`` method to provide the ``sum`` output of the calculation its final value (of type AiiDA integer, as was specified in the ``define`` section of the associated |CalcJob|).
It is important to note here that there is no return statement: the output is provided to the ``self.out`` method instead (any returned value is interpreted as an error signal) as an unstored node, and the engine will be in charge of performing the storing process.

In order to use this ``ArithmeticAddParser`` inside an appropriate |CalcJob| (such as the one described in the :ref:`previous section <how-to:codes:interfacing>`), one needs to add it as a `metadata.options.parser_name` input.

> You can leave it for the user to do so when instantiating the desired calculation plugin, or, if you are developing the calculation plugin, you can add it as the default option in the ``define`` method:

Reviewer 2: This how-to is written for users, so this sentence is confusing.

Author: Changed it, clearer?

In order to use this ``ArithmeticAddParser`` inside an appropriate |CalcJob| (such as the one described in the :ref:`previous section <how-to:codes:interfacing>`), one needs to add it as a `metadata.options.parser_name` input.
You can set a parser as the default option in the |define| method, but note that this choice can be overriden when instantiating the |CalcJob|.

Reviewer 2: Better!

How to run external codes
*************************

To run an external code with AiiDA, you will need to use an appropriate :ref:`calculation plugin <topics:plugins>`.
Its job is to orchestrate the whole process that starts with transforming the input nodes to the input files that the code expects, then copy everything to the machine where the code is located (using a :ref:`transport plugin <how-to:codes:transport>`), run the actual code (using a :ref:`scheduler plugin <how-to:codes:scheduler>`), and finally get back the outputs and, optionally, parse them into queryable AiiDA nodes (using a :ref:`parser plugin <how-to:codes:parser>`).

The following subsections will take you through the process of :ref:`creating the calculation plugin<how-to:codes:interfacing>` and its acompanying :ref:`parser<how-to:codes:parsing>`, and then using these to actually :ref:`run the code<how-to:codes:run>`.

> Other supporting sections are also included.

Reviewer 2: After reviewing this once more, I think the "Design guidelines" from line 144 should go right after this paragraph.

Author: Ok, I can change it. The only caveat is that this are designing guidelines, and although most sections are about designing (calcjobs, parsers, adding caching, schedulers, transports), there are some that are not (setting up computer and running), and furthermore the whole section is called "How to **run** external codes", so it might be weird to read about **design** guidelines in the begining. Please confirm if this is ok.

Reviewer 2: Yeah, let's just go with it for now.

   | Functionality at deeper nesting levels is not considered part of the public API and may change between minor AiiDA releases, breaking your plugin.

.. _how-to:codes:parsing:

Parsing the outputs
===================

The parsing step occurs after the calculation has finished running and all the relevant outputs have been retrieved.
It is an optional step that allows you to extract relevant information from the output files and store it into AiiDA nodes in formats that are easier and quicker to query and analyze.
The way to implement this is by extending the base |Parser| class as you can see below:

.. code-block:: python
   :linenos:

    from aiida.parsers.parser import Parser
    from aiida import orm

    class ArithmeticAddParser(Parser):

        def parse(self, **kwargs):

            output_folder = self.retrieved
            output_filename = self.node.get_attribute('output_filename')

            with output_folder.open(output_filename, 'r') as handle:
                result = int(handle.read())

            self.out('sum', orm.Int(result))

> The first command in the example (line 8) shows the way to get the ``retrieve`` folder that was generated by the associated |CalcJob| and that, by this point, should contain the files included in the ``retrieve_list`` (as specified in the |prepare_for_submission| method).

Reviewer 2: The `| prepare_for_submission|` reference does not resolve properly.
https://88-77234579-gh.circle-artifacts.com/0/html/howto/codes.html#parsing-the-outputs

The second one (line 9) is also accessing a parameter of the |CalcJob|: in this case, the name of the output.
After unloading this information into local variables, these are then used to open said output file located in the obtained retrieved folder and read the single integer value that was written there by the addition |Code| (lines 11 and 12).

Finally, it uses the ``out`` method to provide the ``sum`` output of the calculation its final value (of type AiiDA integer, as was specified in the |define| section of the associated |CalcJob|).
It is important to note here that there is no return statement: the output is provided to the ``self.out`` method instead (any returned value is interpreted as an error signal) as an unstored node, and the engine will be in charge of performing the storing process.

> In order to use this ``ArithmeticAddParser`` inside an appropriate |CalcJob| (such as the one described in the :ref:`previous section <how-to:codes:interfacing>`), one needs to add it as a `metadata.options.parser_name` input.

Reviewer 2: ```suggestion
In order to use this ``ArithmeticAddParser`` inside an appropriate |CalcJob| (such as the one described in the :ref:`previous section <how-to:codes:interfacing>`), one needs to add it as a ``metadata.options.parser_name`` input.
```
Code quotes  must be placed within two backticks for reStructuredText.

You then have to introduce all of these `exit_codes` inside of the |define| method of the |CalcJob| that will use this |Parser| (so, in that sense, when working like this the exit codes need to be "supported" by any calculation that wants to use the |Parser|).

.. code-block:: python

    @classmethod
    def define(cls, spec):
        (...)
        spec.exit_code(300, 'ERROR_NO_RETRIEVED_FOLDER', message='The retrieved folder data node could not be accessed.')
        spec.exit_code(310, 'ERROR_READING_OUTPUT_FILE', message='The output file could not be read from the retrieved folder.')
        spec.exit_code(320, 'ERROR_INVALID_OUTPUT', message='The output file contains invalid output.')

As you can see, for each ``exit_code`` you need to provide an ID number that will be used to identify it, a label you can then use to reference the code in the |parse| method (``self.exit_codes.LABEL``), and a message that will give the user more information on the problem.
The Topics section on :ref:`defining processes <topics:processes:usage:defining>` also contains more information on how to use exit codes.

Design guidelines
-----------------

 * | **Parse what you want to query for.**

>    | Make a list of which information to:

Reviewer 2: This is a list with only one bullet point. Is there supposed to be another bullet point?

Author: The list of all guidelines was only one and I took out the point that was about parsing to put it in the parse section. I thought it would be better to keep the same formet even though that meant having a one item list, but now I tend to agree that its not so good. Anyways, if we go with your other suggestion I will just put all guidelines back together, otherwise I may remove this subsection and just add the content as a paragraph at the endo of the parse section.

Reviewer 2: Let's keep it all together.

     #. parse into the database for querying (:py:class:`~aiida.orm.nodes.data.dict.Dict`, ...)
     #. store in the file repository for safe-keeping (:py:class:`~aiida.orm.nodes.data.singlefile.SinglefileData`, ...)

>      #. leave on the computer where the calculation ran (:py:class:`~aiida.orm.nodes.data.remote.RemoteData`, ...)

Reviewer 2: It's not clear to me what the meaning of the ellipses is here.

Author: So, I didn't write this part so this is just a guess, but I think it is suggesting one posibility of other options? Like, to store queryable data you can do it as a `Dict` but maybe some data would be more convenient to store as an array.

Reviewer 2: Ok, sorry the provenance of those lines got lost. Nevermind then.

`#3990`_

.. |Int| replace:: :py:class:`~aiida.orm.nodes.data.int.Int`
.. |Code| replace:: :py:class:`~aiida.orm.nodes.data.Code`
.. |Parser| replace:: :py:class:`~aiida.parsers.parser.Parser`
.. |parse| replace:: :py:class:`~aiida.parsers.parser.Parser.parse`
.. |folder| replace:: :py:class:`~aiida.common.folders.Folder`
.. |folder.open| replace:: :py:class:`~aiida.common.folders.Folder.open`
.. |CalcJob| replace:: :py:class:`~aiida.engine.processes.calcjobs.calcjob.CalcJob`
.. |CalcInfo| replace:: :py:class:`~aiida.common.CalcInfo`
.. |CodeInfo| replace:: :py:class:`~aiida.common.CodeInfo`
.. |spec| replace:: ``spec``
.. |define| replace:: :py:class:`~aiida.engine.processes.calcjobs.CalcJob.define`

> .. |prepare_for_submission| :py:class:`~aiida.engine.processes.calcjobs.CalcJob.prepare_for_submission`

Reviewer 2: I think sphinx might not like the underscore, but not sure.

Author: Nah, it was the fact that I was not telling it to actually `replace::` the thing =P.
